src/hooks/usePermit2Allowance.test.ts

```diff
@@ -0,0 +1,268 @@
+import { PERMIT2_ADDRESS } from '@uniswap/permit2-sdk'
+import { CurrencyAmount, MaxUint256 } from '@uniswap/sdk-core'
+import { UNIVERSAL_ROUTER_ADDRESS } from '@uniswap/universal-router-sdk'
+import { SupportedChainId } from 'constants/chains'
+import { UNI } from 'constants/tokens'
+import { useAddTransactionInfo, usePendingApproval } from 'hooks/transactions'
+import { usePermitAllowance, useUpdatePermitAllowance } from 'hooks/usePermitAllowance'
+import { useTokenAllowance, useUpdateTokenAllowance } from 'hooks/useTokenAllowance'
+import ms from 'ms.macro'
+import { act, renderHook, waitFor } from 'test'
+
+import usePermit2Allowance, { AllowanceRequired, AllowanceState } from './usePermit2Allowance'
+
+const SPENDER = UNIVERSAL_ROUTER_ADDRESS(SupportedChainId.MAINNET)
+
+const TOKEN = UNI[SupportedChainId.MAINNET]
+const AMOUNT = CurrencyAmount.fromRawAmount(TOKEN, MaxUint256.toString())
+
+jest.mock('hooks/useTokenAllowance')
+jest.mock('hooks/usePermitAllowance')
+jest.mock('hooks/transactions')
+const mockUseTokenAllowance = useTokenAllowance as jest.Mock
+const mockUseUpdateTokenAllowance = useUpdateTokenAllowance as jest.Mock
+const mockUsePermitAllowance = usePermitAllowance as jest.Mock
+const mockUseUpdatePermitAllowance = useUpdatePermitAllowance as jest.Mock
+const mockUseAddTransactionInfo = useAddTransactionInfo as jest.Mock
+const mockUsePendingApproval = usePendingApproval as jest.Mock
+
+const addressMatcher = expect.stringMatching(new RegExp(hardhat.account.address, 'i'))
+
+beforeEach(jest.resetAllMocks)
+
+describe('usePermit2Allowance', () => {
+  beforeEach(() => {
+    mockUseTokenAllowance.mockReturnValue({ tokenAllowance: undefined })
+    mockUsePermitAllowance.mockReturnValue({ permitAllowance: undefined })
+  })
+
+  describe('with no token', () => {
+    it('defaults to allowed', () => {
+      const { result } = renderHook(() => usePermit2Allowance())
+      expect(result.current).toMatchObject({ state: AllowanceState.ALLOWED })
+    })
+  })
+
+  it('loads', async () => {
+    const { result } = renderHook(() => usePermit2Allowance(AMOUNT, SPENDER))
+    await waitFor(() => expect(useTokenAllowance).toHaveBeenCalledWith(TOKEN, addressMatcher, PERMIT2_ADDRESS))
+    expect(usePermitAllowance).toHaveBeenCalledWith(TOKEN, addressMatcher, SPENDER)
+    expect(result.current).toMatchObject({ state: AllowanceState.LOADING })
+  })
+
+  describe('approves', () => {
```

Comment:
better desc

Reply:
should each of these should be separate tests? otherwise the entire test will fail vs seeing exactly what went wrong.

Reply:
these _are_ the separate tests. they track the different starting states for approval, and how that settles.
they all deal with intermediate state, so I don't think I can break them down any further.

Reply:
I split these a bit more, but they still test intermediate states, so some will remain longer.